staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -137,159 +208,90 @@ func New(opts OIDCOptions) (*OIDCAuthenticator, error) {
 
-	// Acquire lock, then recheck initialization.
-	a.mu.Lock()
-	defer a.mu.Unlock()
-	if client := a.oidcClient.Load(); client != nil {
-		return client.(*oidc.Client), nil
-	}
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx = oidc.ClientContext(ctx, client)
 
-	// Try to initialize client.
-	providerConfig, err := oidc.FetchProviderConfig(a.httpClient, a.issuerURL)
-	if err != nil {
-		glog.Errorf("oidc authenticator: failed to fetch provider discovery data: %v", err)
-		return nil, fmt.Errorf("fetch provider config: %v", err)
+	authenticator := &Authenticator{
+		issuerURL:      opts.IssuerURL,
+		usernameClaim:  opts.UsernameClaim,
+		usernamePrefix: opts.UsernamePrefix,
+		groupsClaim:    opts.GroupsClaim,
+		groupsPrefix:   opts.GroupsPrefix,
+		ctx:            ctx,
+		cancel:         cancel,
 	}
 
-	clientConfig := oidc.ClientConfig{
-		HTTPClient:     a.httpClient,
-		Credentials:    oidc.ClientCredentials{ID: a.trustedClientID},
-		ProviderConfig: providerConfig,
+	now := opts.now
+	if now == nil {
+		now = time.Now
 	}
 
-	client, err := oidc.NewClient(clientConfig)
-	if err != nil {
-		glog.Errorf("oidc authenticator: failed to create client: %v", err)
-		return nil, fmt.Errorf("create client: %v", err)
+	verifierConfig := &oidc.Config{
+		ClientID:             opts.ClientID,
+		SupportedSigningAlgs: supportedSigningAlgs,
+		Now:                  now,
 	}
 
-	// SyncProviderConfig will start a goroutine to periodically synchronize the provider config.
-	// The synchronization interval is set by the expiration length of the config, and has a minimum
-	// and maximum threshold.
-	stop := client.SyncProviderConfig(a.issuerURL)
-	a.oidcClient.Store(client)
-	a.close = func() {
-		// This assumes the stop is an unbuffered channel.
-		// So instead of closing the channel, we send am empty struct here.
-		// This guarantees that when this function returns, there is no flying requests,
-		// because a send to an unbuffered channel happens after the receive from the channel.
-		stop <- struct{}{}
-	}
-	return client, nil
+	initVerifier(authenticator, verifierConfig)
+	return authenticator, nil
 }
 
-// AuthenticateToken decodes and verifies an ID Token using the OIDC client, if the verification succeeds,
-// then it will extract the user info from the JWT claims.
-func (a *OIDCAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
-	jwt, err := jose.ParseJWT(value)
-	if err != nil {
-		return nil, false, err
+func (a *Authenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+	ctx := context.Background()
+	verifier, ok := a.idTokenVerifier()
+	if !ok {
+		return nil, false, fmt.Errorf("oidc: authenticator not initialized")
 	}
 
-	client, err := a.client()
-	if err != nil {
-		return nil, false, err
-	}
-	if err := client.VerifyJWT(jwt); err != nil {
-		return nil, false, err
-	}
-	claims, err := jwt.Claims()
+	idToken, err := verifier.Verify(ctx, token)
```

Comment:
can be a follow-up, but might be good to consider doing something similar to https://github.com/kubernetes/kubernetes/pull/58791 to avoid noise in logs when this is combined with other token auth methods